Add encoding='utf-8' to open() in setup.py

9999years opened the pull request:
Currently, machines which default to CP-1252 error when reading CHANGELOG.rst (there are a few names with non-Latin characters or diacritics); `UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2863: character maps to <undefined>`. 

By adding an `encoding='utf-8'` keyword argument to the two `open()` calls in `setup.py`'s `get_full_description`, we can prevent this error.

This change is backwards-compatible, so only a patch version bump is needed.

asfaltboy commented:
Hi @9999years thanks for contributing!

[Travis says][1] `encoding` is an invalid keyword argument for `open` 🤷‍♂️ . Guess we should either decode binary input explicitly or use [`io.open()`][2] ([ref SO answer][3])

[1]: https://travis-ci.org/modlinltd/django-advanced-filters/jobs/438794196
[2]: https://docs.python.org/2/library/io.html#io.open
[3]: https://stackoverflow.com/a/10975371

9999years commented:
Changed `open` to `io.open` and squashed up the commits; all green now! 🎉

asfaltboy commented:
Thanks again. Hmm, I'd like to test out the fix, could you help me understand how I can reproduce the issue?

I tried a few variations of the following, but could not reproduce:

```
env LC_ALL="en_US.CP1252" python setup.py sdist
```

9999years commented:
Well, to be frank I’m not entirely sure how it’d be reproduced on Linux. I’ll look into it, though.

Here’s the full error, though:

```
pip install django-advanced-filters
Collecting django-advanced-filters
  Downloading https://files.pythonhosted.org/packages/45/68/98aed982c7a56eb47dfed7cf47703c533b7bdc4ea65f85bae7fbf98c48eb/django-advanced-filters-1.1.0.tar.gz (72kB)
    100% |████████████████████████████████| 81kB 2.6MB/s
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\arvensis\AppData\Local\Temp\pip-install-zt4kcl1x\django-advanced-filters\setup.py", line 62, in <module>
        long_description=get_full_description(),
      File "C:\Users\arvensis\AppData\Local\Temp\pip-install-zt4kcl1x\django-advanced-filters\setup.py", line 40, in get_full_description
        CHANGELOG = changelog.read()
      File "c:\python37\lib\encodings\cp1252.py", line 23, in decode
        return codecs.charmap_decode(input,self.errors,decoding_table)[0]
    UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2863: character maps to <undefined>

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\Users\arvensis\AppData\Local\Temp\pip-install-zt4kcl1x\django-advanced-filters\
```

9999years commented:
This seems to be pretty intertwined with the frustrating way Windows CPython interacts with encodings. See:

```
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale
>>> locale.setlocale(locale.LC_ALL, ('en_us', 'UTF-8'))
'en_US.UTF-8'
>>> print('Locale:', locale.getlocale())
Locale: ('en_US', 'UTF-8')
>>> print('Encoding:', locale.getpreferredencoding())
Encoding: cp1252
```

Even after we explicitly set the locale to use UTF-8, our setting is abandoned (and the error persists).

See also in [`_bootlocale.py`](https://github.com/python/cpython/blob/master/Lib/_bootlocale.py#L10) how the Windows implementation of `getpreferredencoding` uses the Windows implementation in [`_localemodule.c`](https://github.com/python/cpython/blob/master/Modules/_localemodule.c#L320) which itself delegates to the winapi [`GetACP`](https://docs.microsoft.com/en-us/windows/desktop/api/winnls/nf-winnls-getacp) function which returns a codepage number.

What I’m getting at is the locale behavior is a total platform-dependent mess.

Anyways, it seems to boil down to the fact that `en_US.CP1252` is not supported on most Linux systems (check `locale -a` to confirm, though) so that setting is probably abandoned by the C `setlocale` call. The easiest way to recreate this is running it on a Windows machine — I’ve been digging through the Python source trying to find another way but it’s just not worth the effort.

The fix makes `setup.py` run without errors on Windows (and, critically, consistently on any platform with a default locale that’s not UTF-8). Run my branch on Windows and it runs fine, while `master` crashes.

```
 ⇒ git checkout patch-1; pip install -e .
Switched to branch 'patch-1'
Your branch is up to date with 'origin/patch-1'.
Obtaining file:///C:/Users/arvensis/Documents/projects/django-advanced-filters
Requirement already satisfied: django-braces<2,>=1.4.0 in c:\python37\lib\site-packages (from django-advanced-filters==1.1.0) (1.13.0)
Requirement already satisfied: simplejson<4,>=3.6.5 in c:\python37\lib\site-packages (from django-advanced-filters==1.1.0) (3.16.0)
Installing collected packages: django-advanced-filters
  Found existing installation: django-advanced-filters 1.1.0
    Uninstalling django-advanced-filters-1.1.0:
      Successfully uninstalled django-advanced-filters-1.1.0
  Running setup.py develop for django-advanced-filters
Successfully installed django-advanced-filters
~\Documents\projects\django-advanced-filters
 ⇒ git checkout master; pip install -e .
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
Obtaining file:///C:/Users/arvensis/Documents/projects/django-advanced-filters
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\arvensis\Documents\projects\django-advanced-filters\setup.py", line 62, in <module>
        long_description=get_full_description(),
      File "C:\Users\arvensis\Documents\projects\django-advanced-filters\setup.py", line 40, in get_full_description
        CHANGELOG = changelog.read()
      File "c:\python37\lib\encodings\cp1252.py", line 23, in decode
        return codecs.charmap_decode(input,self.errors,decoding_table)[0]
    UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2863: character maps to <undefined>

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\Users\arvensis\Documents\projects\django-advanced-filters\
```

maingoh commented:
Related to #95. It is quite easy to reproduce with a docker and no locale set.

asfaltboy commented:
If it works for everyone, I'm going to go ahead and merge it. Sorry for delay

maingoh commented:
You can try by yourself with this dockerfile:
```
FROM ubuntu:18.04                                                                                                                                    
RUN apt-get update && apt-get install -y --no-install-recommends git python3-dev python3-pip python3-setuptools python3-wheel                                                                                                             
# RUN pip3 install django-advanced-filters==1.1.0 # This crash
RUN pip3 install git+https://github.com/9999years/django-advanced-filters.git@patch-1 # Working
```
